### PR TITLE
AIRSpecializeChannelWrapAndStride: Hoist loop-invariant pure ops for dominance

### PR DESCRIPTION
**Pull request by erwei-xilinx**
Below shows the standard case that the pass has been supporting, where the `affine.apply` can get folded into the `scf.for` loop bounds:
```
    scf.for %arg4 = %c0 to %c8 step %c1 {
      %0 = affine.apply #map()[%arg4]
      %1 = air.channel.put async @channel_21[] (%arg0[%0] [%c8] [%c1]) : (memref<128x256xi32>)
    }
```
However, user-provided code could come in the following form too:
```
    scf.for %arg4 = %c0 to %c8 step %c1 {
      %0 = affine.apply #map()[%arg1]
      %1 = air.channel.put async @channel_21[] (%arg0[%0] [%c8] [%c1]) : (memref<128x256xi32>)
    }
```
Where the `affine.apply` cannot fold into the loop due to `%arg1` not being the loop induction variable. In such case, the pass should hoist the `affine.apply` out of the for loop in order to fix value domination, since `affine.apply` is a pure op that doesn't touch memory and can therefore hoist safely.

### CODE REVIEW

File: mlir/lib/Util/Util.cpp [hunk 1]
```diff
@@ -1577,8 +1577,27 @@ air::evaluateConstantsInMap(AffineMap map,
...
-  // auto newmap = map.replace(getAffineSymbolExpr(0, ctx), c, 0, 1);
   output = simplifyAffineMap(newmap).getSingleConstantResult();
   return output;
 }
+
+// Extend the lookupOrDefault method to operate on a vector of values.
+Value air::lookupOrDefaultRange(Value v, IRMapping &remap) {
+  return remap.lookupOrDefault(v);
+}
+SmallVector<Value> air::lookupOrDefaultRange(SmallVector<Value> vec,
```
Comment (fifield) [inline mlir/lib/Util/Util.cpp:1588]:
```suggestion
SmallVector<Value> air::lookupOrDefaultRange(SmallVector<Value> &vec,
```

File: mlir/lib/Transform/AIRDependencyScheduleOpt.cpp [hunk 2]
```diff
@@ -2087,13 +2137,13 @@ struct AIRUnrollAffineForIntoBDChain
...
-        else if (auto exec = dyn_cast<air::ExecuteOp>(o)) {
-          auto childOp = exec.getChildOp();
-          if (childOp && isPure(childOp))
-            continue;
-        }
+        // else if (auto exec = dyn_cast<air::ExecuteOp>(o)) {
+        //   auto childOp = exec.getChildOp();
+        //   if (childOp && isPure(childOp))
+        //     continue;
+        // }
```
Comment (fifield) [inline mlir/lib/Transform/AIRDependencyScheduleOpt.cpp:2146]:
```suggestion
```

File: mlir/lib/Transform/AIRDependencyScheduleOpt.cpp [hunk 3]
```diff
@@ -2024,6 +2048,37 @@ struct AIRSpecializeChannelWrapAndStrideInScfFor
   }
 
 private:
+  // Clone backward slices of a list of values, opers.
+  SmallVector<Operation *> cloneDefiningOpsInRegion(OpBuilder builder,
+                                                    Region *region,
+                                                    SmallVector<Value> opers,
```
Comment (fifield) [inline mlir/lib/Transform/AIRDependencyScheduleOpt.cpp:2054]:
```suggestion
                                                    SmallVector<Value> &opers,
```

File: mlir/lib/Transform/AIRDependencyScheduleOpt.cpp [hunk 4]
```diff
@@ -2024,6 +2048,37 @@ struct AIRSpecializeChannelWrapAndStrideInScfFor
...
+      if (!region->isAncestor(operandDefOp->getParentRegion()))
+        continue;
+      assert(air::isPure(
+          operandDefOp)); // Pure ops are safe to hoist out of loops.
+      // Get backward slices
+      SetVector<Operation *> operandBS;
+      getBackwardSlice(operandDefOp, &operandBS, bsOptions);
+      for (auto b : operandBS)
+        if (air::isPure(b))
+          backwardSlices.insert(b);
```
Comment (fifield) [inline mlir/lib/Transform/AIRDependencyScheduleOpt.cpp:2075]: We assert that the operands are pure, but not the operands' operands?

File: mlir/lib/Util/Util.cpp [hunk 5]
```diff
@@ -1577,8 +1577,27 @@ air::evaluateConstantsInMap(AffineMap map,
...
+    output.push_back(remap.lookupOrDefault(v));
+  }
+  return output;
+}
+
+// Extend isPure method to operate on air.execute.
+bool air::isPure(Operation *op) {
+  bool result = mlir::isPure(op);
+  if (auto execOp = dyn_cast<air::ExecuteOp>(op))
+    result = mlir::isPure(execOp.getChildOp());
```
Comment (fifield) [inline mlir/lib/Util/Util.cpp:1601]: Is air.execute always assumed to have only one op inside? This is not checked by the verifier and we should relax it to not be the case.

File: mlir/lib/Util/Util.cpp (hunk 1 above)
Comment (erwei-xilinx) [reply mlir/lib/Util/Util.cpp]: Make sense. Updated.

File: mlir/lib/Transform/AIRDependencyScheduleOpt.cpp (hunk 2 above)
Comment (erwei-xilinx) [reply mlir/lib/Transform/AIRDependencyScheduleOpt.cpp]: Oops, deleted.

File: mlir/lib/Transform/AIRDependencyScheduleOpt.cpp (hunk 3 above)
Comment (erwei-xilinx) [reply mlir/lib/Transform/AIRDependencyScheduleOpt.cpp]: Updated.

File: mlir/lib/Transform/AIRDependencyScheduleOpt.cpp (hunk 4 above)
Comment (erwei-xilinx) [reply mlir/lib/Transform/AIRDependencyScheduleOpt.cpp]: Assertion added.

File: mlir/lib/Util/Util.cpp (hunk 5 above)
Comment (erwei-xilinx) [reply mlir/lib/Util/Util.cpp:1609]: Originally, yes. But I agree with you that we should relax this constraint. Let me create a separate PR for it.

Comment (erwei-xilinx) [reply mlir/lib/Util/Util.cpp:1609]: PR created: https://github.com/Xilinx/mlir-air/pull/773

Comment (erwei-xilinx) [reply mlir/lib/Util/Util.cpp:1609]: PR created: https://github.com/Xilinx/mlir-air/pull/773

**Pull request merged**